coop_shift/model/shift_mail_registration.py

```diff
@@ -35,20 +35,24 @@ class ShiftMailRegistration(models.Model):
         'shift.registration', 'Attendee', required=True, ondelete='cascade')
     mail_ignored = fields.Boolean('Ignored', default=False)
 
-    @api.one
+    @api.multi
     def execute(self):
-        today = datetime.strftime(fields.datetime.now(),
-                                  tools.DEFAULT_SERVER_DATETIME_FORMAT)
-        # send email for user if shift hasn't started
-        if self.registration_id.shift_id.date_begin >= today:
-            # when users is in vacation, no email is sent to them.
-            if self.registration_id.partner_id.working_state in ['exempted',
-                                                                 'vacation']:
-                return
-            if self.registration_id.shift_id.is_on_holiday:
-                return
-            return super(ShiftMailRegistration, self).execute()
-        else:
-            # other case which not sending mail, marked ignore=True
-            # they won't be sent ever.
-            return self.write({'mail_ignored': True})
+        for rec in self:
+            today = datetime.strftime(fields.datetime.now(),
+                                      tools.DEFAULT_SERVER_DATETIME_FORMAT)
+            today = datetime.strptime(today,
+                                      tools.DEFAULT_SERVER_DATETIME_FORMAT)
+            # send email for user if shift hasn't started
+            if rec.registration_id.shift_id.date_begin >= today:
+                # when users is in vacation, no email is sent to them.
+                if rec.registration_id.partner_id.working_state in [
+                        'exempted', 'vacation']:
+                    return
+                # TODO: remove 'is_on_holiday' is not found in coop_shift
+                # if rec.registration_id.shift_id.is_on_holiday:
+                #     return
+                return super(ShiftMailRegistration, rec).execute()
+            else:
+                # other case which not sending mail, marked ignore=True
+                # they won't be sent ever.
+                return rec.write({'mail_ignored': True})
```

Comment:
This `for rec in self` loop needs to be refactored.
You can't `return` inside the loop, it'll stop it.

If you call this function on an actual recordset with multiple records, it'll execute only the first one!


On way of refactoring this could be filtering the `self` recordset based on your conditions, and then use set operations to split the recordset and call the corresponding methods.

```py
records_to_execute = self.filtered(lambda rec: ...)
records_to_ignore = self - records_to_execute
records_to_skip = records_to_execute.filtered(lambda rec: rec.registration_id.partner_id.working_state in [..])
records_to_execute -= records_to_skip

if records_to_ignore:
    records_to_ignore.write({'mail_ignored': True})

return super(ShiftMailRegistration, records_to_execute).execute()
```